### ISSUE TITLE
QGIS Plugin and Xarray Extension

### ISSUE THREAD

**Issue by Huite** [pyvista/pyvista#629]
Hello all,

I'd like to start with congratulating you with this wonderful package. PyVista has allowed to me *finally* setup 3D visualizations quickly and intuitively through Python, and I'm immensely grateful!

There are, in my humble view, very large gains in linking PyVista with two other open source packages: 
* [xarray](https://github.com/pydata/xarray)
* [QGIS](https://github.com/qgis/QGIS)

These are both rather popular (and for good reason), and I think both would be significantly strengthened by convenient 3D visualization as provided by PyVista. Additionally, both provide convenient and unintrusive ways of extending functionality. @JiaweiZhuang has also mentioned xarray here before.

As @agramfort has mentioned in #422, many alternatives appear too fragile or complex, and PyVista is especially appealing because of its simple API and clear scope.

### Extending xarray
Xarray provides an extension interface: https://xarray.pydata.org/en/stable/internals.html#extending-xarray

See an example here: https://github.com/corteva/rioxarray

We could have a package like `pyvistaxarray` (preferably something catchier), which would roughly look like this:

```python
import xarray as xr
import pyvistaxarray

# open netCDF with dims (time, z, y, x)
da = xr.open_dataarray("example.nc")  
# get a single time and z slice and plot in 2D with matplotlib
da.sel(time="2001-01-01").isel(z=0).plot()
# get a single time slice and plot in 3D with pyvista
da.sel(time="2001-01-01").pyvista.plot()   
```

In this case, the plot function collects the xarray coordinates and generates a PyVista Mesh.
Xarray is very powerful in terms of indexing, slicing, conditional selection, reductions over specific dimensions, etc. For example, to compute the first interface in the z dimension:

```python
interface = da["z"].where(da > 5.0).max("z")
interface.pyvista.plot()  # Now draw the interface as a surface in 3D
```

I've written a few functions (for rectilinear grids with potentially a lot of nodata values) here:
https://gitlab.com/deltares/imod/imod-python/-/blob/master/imod/visualize/pyvista.py

### Extending QGIS
QGIS provides a plugin structure:
https://plugins.qgis.org/

It's not quite as straightforward as extending xarray, but I've spent a day or two looking at how these plugins work and I'm convinced it's thoroughly manageable! I hacked a prototype together in a four hours or so, yesterday:

![qgis-pyvista](https://user-images.githubusercontent.com/13662783/75114361-9ded3980-5655-11ea-84d9-a08e9f998fb3.PNG)

Left the QGIS window, top right the layer selection, bottom right the BackgroundPlotter.

Currently, it's a few components;
* A table to select the QGIS layers
* Something that fetches the QGIS color styling
* Get the viewing extent, use GDAL to get a numpy array
* Generate a mesh, and plot it in the BackgroundPlotter

It's all Qt-based, and so far, it works together without a hitch. I'm looking at https://github.com/pyvista/pyvista-gui with great interest, of course.

# The bottom line
I'm going to start pushing for the above in my organisation, but before we make any decisions, I'd like to know whether you're aware of any existing initiatives. As hydrologists, we work a great deal with multidimensional and geospatial data and I reckon that's also the case for people in e.g. https://github.com/simpeg, so maybe somebody's already spent some time on this.

**Comment by banesullivan** [comment]
This is really, really awesome investigative work, @Huite! Thanks for sharing!!

At the moment there are no existing efforts for these interfaces though I would be personally interested in expanding both (and I think I could get a few other folks on board down the road). At the moment, I'm burnt out and about to take a month or two offline but I would be happy to help collaborate on both of those interfaces moving forward!

**Comment by Huite** [comment]
Great, thanks for the info! I'll keep you updated.

Have a good break!

**Comment by GGDRriedel** [comment]
I am hugely interested in this but I am more interested in Bane taking a break

**Comment by akaszynski** [comment]
@banesullivan is quite busy with his day job (and so am I), so any additional features will have to be added by the community.  I think @Huite has started some work according to https://gitlab.com/deltares/imod/imod-python/-/blob/master/imod/visualize/pyvista.py, and we can expand on that by adding some of that functionality as as a sub-module within `pyvista`.  I can review a PR and provide guidance, but I won't be able to write anything since `pyvista` doesn't pay the bills.

**Comment by banesullivan** [comment]
FYI for the xarray integration: https://github.com/pydata/xarray/issues/4470 This is something we are looking into

**Comment by dephora** [comment]
@Huite - Does QGIS not crash upon closing when importing pyvista (not even using it, just simply importing it) - see pyvista/pyvistaqt#69.

I don't want to hijack this but I'm in the process of trying to resolve this.

**Comment by Huite** [comment]
Hi @dephora, I just tried it within my QGIS 3.18, seems to still work like a charm:

![image](https://user-images.githubusercontent.com/13662783/133464353-92aa61a6-445f-49b0-bb60-84e5e7cc188f.png)

I've tried:

* Importing pyvista
* Importing pyvistaqt
* Opening pyvistaqt window
* Closing pyvistaqt window
* Opening pyvistaqt window again

No trouble so far.

I'm on Windows, just installed pyvista & pyvistaqt fresh.

I have had a number of other issues running things in the QGIS Python interpreter. For example, I can't get Numba to JIT-compile stuff in QGIS. In a number of cases, these issues seem related to environmental variables that are set for QGIS. I can give some ideas in the linked issue.

**Comment by dephora** [comment]
@Huite  - Thanks for the quick response! So everything you have listed does work on my end in various conda environments with the conda-forge QGIS & pyvista or through the standalone QGIS with a pip install of pyvista. The issue arises only when you actually close QGIS at which point it crashes (as seen in the other repo).

**Comment by Huite** [comment]
Uhhh yes, looks like my reading comprehension wasn't up to scratch after this day of work. After finishing dinner now and only now closing QGIS it indeed crashes 😅

**Comment by dephora** [comment]
No worries, your alternate solution in the other repo is good information. That being said, vedo (which uses VTK) works without issue, so I imagine there is a fix to avoid that route.

**Comment by banesullivan** [comment]
Apologies for bringing an old issue to life, but I wanted to point folks to `pyvista-xarray` for an xarray interface: https://github.com/pyvista/pyvista-xarray

I'm also going to close this issue due to staleness for the QGIS aspect

**Comment by GGDRriedel** [comment]
@Huite Whatever happened to this? It would revolutionize Geophysical data display and reporting

**Comment by GGDRriedel** [comment]
Would you be open to go open source with this? I think the geophysical community would have a HUGE HUGE interest in this, maybe we could even get some developers interested if we get a small prototype working.

**Comment by GGDRriedel** [comment]
Also yes, I am talking about the QGIS integration :) 

We basically have two types of data:  Crooked profiles (aka 2.5 D) and I got texture generation as well as display of that working through pyvista, currently using Paraview for display though, and 3D data that can be seen as irregular voxels. 
Not that much of a problem in regards to data since we can just use the smallest voxel representation and multiples of it for the 3D stuff. 


My problem is more that I am not very fluent in the inner workings of QGIS, I am just scientific Python guy . Tbh. I would be more than happy to see your initial example up there and try to work off that whenever I get the time

**Comment by GGDRriedel** [comment]
Look!

https://www.lutraconsulting.co.uk/crowdfunding/qgis-3d-for-open-source-digital-twins

**Comment by Huite** [comment]
Hi @GGDRriedel 

Not to dampen the enthusiasm for QGIS developments, but: in my view these (and past) QGIS 3D developments have a different focus. More so on a particular kind of 3D data, points clouds, the built environment, etc. Less so on the simulation meshes and the like that the VTK stack facillitates. I reckon the linked spec is useful context: https://github.com/Esri/i3s-spec

This makes sense for QGIS as a general geospatial toolkit, the surface of the earth is relevant to all kinds of users; the subsurface is a smaller niche.

Anyway, I'll try to find some time to revive and update the proof-of-concept thing and stuff in into a Github repo.
